src/wallet/wallet.cpp

```diff
@@ -4204,3 +4204,61 @@ void CWallet::SetActiveScriptPubKeyMan(uint256 id, OutputType type, bool interna
     spk_man->SetType(type, internal);
     spk_mans[type] = spk_man;
 }
+
+bool CWallet::HasWalletDescriptor(const WalletDescriptor& desc) const
+{
+    if (!IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS)) {
+        return false;
+    }
+
+    for (auto spk_man_pair : m_spk_managers) {
+        // Try to downcast to DescriptorScriptPubKeyMan then check if the descriptors match
+        std::shared_ptr<DescriptorScriptPubKeyMan> spk_manager = std::dynamic_pointer_cast<DescriptorScriptPubKeyMan>(spk_man_pair.second);
+        if (spk_manager != nullptr && spk_manager->HasWalletDescriptor(desc)) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+std::shared_ptr<ScriptPubKeyMan> CWallet::AddWalletDescriptor(WalletDescriptor& desc, const FlatSigningProvider& signing_provider, const std::string& label, Optional<OutputType>& parsed_output_type)
+{
+    if (!IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS)) {
+        return nullptr;
+    }
+
+    auto spk_manager = std::shared_ptr<DescriptorScriptPubKeyMan>(new DescriptorScriptPubKeyMan(std::bind(&CWallet::IsWalletFlagSet, this, std::placeholders::_1), std::bind(&CWallet::SetWalletFlag, this, std::placeholders::_1), std::bind(&CWallet::UnsetWalletFlagWithDB, this, std::placeholders::_1, std::placeholders::_2), std::bind(&CWallet::CanSupportFeature, this, std::placeholders::_1), std::bind(&CWallet::GetDisplayName, this), std::bind(&CWallet::SetMinVersion, this, std::placeholders::_1, std::placeholders::_2, std::placeholders::_3), database, desc));
+
+    // Top up key pool, the manager will generate new scriptPubKeys internally
+    spk_manager->TopUp();
+
+    // Determine the output type by peeking into the first scriptPubKey (that we just generated with TopUp())
+    auto script_pub_keys = spk_manager->GetScriptPubKeys();
+    if (script_pub_keys.empty()) {
+        return nullptr;
+    }
+    const CScript& script = script_pub_keys.at(0);
+    parsed_output_type = DetermineOutputType(script, signing_provider);
+
+    // Apply the label if necessary
+    // (This should apply to only one scriptPubKey, as we disable labels for ranged descriptors)
+    if (!label.empty()) {
```

Comment:
We should still apply an empty label for every `scriptPubKey`, so this check is not needed.